core/src/main/java/com/datastax/oss/driver/internal/core/pool/ChannelSet.java

```diff
@@ -83,18 +94,27 @@ DriverChannel next() {
       case 0:
         return null;
       case 1:
-        return snapshot[0];
+        DriverChannel onlyChannel = snapshot[0];
```

Comment:
question not related to this PR: the `snapshot` is only a reference to the array, so if something in the `private volatile DriverChannel[] channels;` will change it will be reflected in the `snapshot` - so to me it seems like not a snapshot (it would need to be an immutable copy to be a real snapshot) Is it intended? Isn't it causing us some problems?

Reply:
Mutation operations (`add()` and `remove()`) always replace the whole array, they don't switch elements in place. So `snapshot` is guaranteed to represent a consistent view of the set at some point in recent time.

The worst that can happen is that `add()` or `remove()` is called concurrently right after we take the snapshot.
- add: we're missing a new channel that just got added, no big deal (it's as if we had done the selection a few nanoseconds before)
- remove: the stale snapshot might contain a channel that is now closed. At the very worst we select it, the client will run into an error when it tries to write, and try the next node.

Reply:
thanks for the explanation